ext/backtracie_native_extension/backtracie.c

```diff
@@ -16,105 +16,133 @@
-static VALUE cfunc_frame_to_location(raw_location *the_location, raw_location *last_ruby_location);
-static VALUE frame_from_location(raw_location *the_location);
-static VALUE qualified_method_name_for_block(raw_location *the_location);
-static VALUE qualified_method_name_from_self(raw_location *the_location);
-static bool is_self_class_singleton(raw_location *the_location);
-static bool is_defined_class_a_refinement(raw_location *the_location);
-static VALUE debug_raw_location(raw_location *the_location);
+static VALUE collect_backtrace_locations(VALUE self, VALUE thread,
+                                         int ignored_stack_top_frames);
+inline static VALUE new_location(VALUE absolute_path, VALUE base_label,
+                                 VALUE label, VALUE lineno, VALUE path,
+                                 VALUE qualified_method_name, VALUE debug);
+static VALUE frame_to_location(const raw_location *raw_loc,
+                               const raw_location *prev_ruby_loc);
+static VALUE debug_raw_location(const raw_location *the_location);
 static VALUE debug_frame(VALUE frame);
-static VALUE cfunc_function_info(raw_location *the_location);
+static VALUE cfunc_function_info(const raw_location *the_location);
 static inline VALUE to_boolean(bool value);
 
+BACKTRACIE_API
 void Init_backtracie_native_extension(void) {
-  main_object_instance = rb_funcall(rb_const_get(rb_cObject, rb_intern("TOPLEVEL_BINDING")), rb_intern("eval"), 1, rb_str_new2("self"));
+  backtracie_main_object_instance =
+      rb_funcall(rb_const_get(rb_cObject, rb_intern("TOPLEVEL_BINDING")),
+                 rb_intern("eval"), 1, rb_str_new2("self"));
   ensure_object_is_thread_id = rb_intern("ensure_object_is_thread");
   to_s_id = rb_intern("to_s");
 
   backtracie_module = rb_const_get(rb_cObject, rb_intern("Backtracie"));
   rb_global_variable(&backtracie_module);
 
-  rb_define_module_function(backtracie_module, "backtrace_locations", primitive_backtrace_locations, 1);
+  rb_define_module_function(backtracie_module, "backtrace_locations",
+                            primitive_backtrace_locations, 1);
 
-  backtracie_location_class = rb_const_get(backtracie_module, rb_intern("Location"));
+  backtracie_location_class =
+      rb_const_get(backtracie_module, rb_intern("Location"));
   rb_global_variable(&backtracie_location_class);
 
-  VALUE backtracie_primitive_module = rb_define_module_under(backtracie_module, "Primitive");
+  VALUE backtracie_primitive_module =
+      rb_define_module_under(backtracie_module, "Primitive");
+
+  rb_define_module_function(backtracie_primitive_module, "caller_locations",
+                            primitive_caller_locations, 0);
 
-  rb_define_module_function(backtracie_primitive_module, "caller_locations", primitive_caller_locations, 0);
+  backtracie_frame_wrapper_class =
+      rb_define_class_under(backtracie_module, "FrameWrapper", rb_cObject);
```

Comment:
Right, my reading is that would stop anybody from instantiating this in Ruby at all, right? (and thus, stop it being instantiated as T_OBJECT rather than T_DATA)

Reply:
Yeap, exactly. `FrameWrapper.new` would fail because `FrameWrapper.allocate` does not exist. (There's actually an alternative option, using a [`rb_define_alloc_func` to define our own custom alloc function](https://github.com/DataDog/dd-trace-rb/blob/master/ext/ddtrace_profiling_native_extension/stack_recorder.c#L41), but I don't think we need it in this case, since we don't expect to create new instances from Ruby)